<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><title>Python: module run_Wochenende</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
</head><body bgcolor="#f0f0f8">

<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="heading">
<tr bgcolor="#7799ee">
<td valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial">&nbsp;<br><big><big><strong>run_Wochenende</strong></big></big></font></td
><td align=right valign=bottom
><font color="#ffffff" face="helvetica, arial"><a href=".">index</a><br><a href="file:/Users/tobi/Documents/Wochenende/run_Wochenende.py">/Users/tobi/Documents/Wochenende/run_Wochenende.py</a></font></td></tr></table>
    <p><tt>Wochenende:&nbsp;A&nbsp;whole&nbsp;genome/metagenome&nbsp;analysis&nbsp;pipeline&nbsp;in&nbsp;Python3&nbsp;(2018-2020)<br>
Author:&nbsp;Tobias&nbsp;Scheithauer<br>
Author:&nbsp;Dr.&nbsp;Colin&nbsp;Davenport<br>
Author:&nbsp;Fabian&nbsp;Friedrich<br>
&nbsp;<br>
&nbsp;<br>
Changelog<br>
1.8.3&nbsp;add&nbsp;ecoli&nbsp;ref<br>
1.8.2&nbsp;Check&nbsp;number&nbsp;of&nbsp;uncommented&nbsp;lines&nbsp;=&nbsp;1&nbsp;in&nbsp;run_Wochenende_SLURM.sh&nbsp;from&nbsp;start&nbsp;script&nbsp;runbatch_sbatch_Wochenende.sh.<br>
1.8.1&nbsp;Move&nbsp;mq30&nbsp;filter&nbsp;(makes&nbsp;big&nbsp;changes)&nbsp;to&nbsp;after&nbsp;mm&nbsp;and&nbsp;duplicate&nbsp;filtering<br>
1.8.0&nbsp;Add&nbsp;name&nbsp;sorting&nbsp;and&nbsp;fixmates&nbsp;for&nbsp;PE&nbsp;reads<br>
1.7.8&nbsp;Test&nbsp;samtools&nbsp;markdup&nbsp;as&nbsp;replacement&nbsp;for&nbsp;sambamba&nbsp;markdup&nbsp;because&nbsp;of&nbsp;16k&nbsp;max&nbsp;ref&nbsp;seqs&nbsp;problem<br>
1.7.7&nbsp;update&nbsp;tests&nbsp;after&nbsp;moving&nbsp;to&nbsp;subdir<br>
1.7.6&nbsp;add&nbsp;2020_09&nbsp;massive&nbsp;reference&nbsp;with&nbsp;all&nbsp;bacterial&nbsp;strains.<br>
1.7.5&nbsp;add&nbsp;trim_galore&nbsp;trimmer&nbsp;for&nbsp;nextera&nbsp;(SE&nbsp;reads&nbsp;only&nbsp;so&nbsp;far)<br>
1.7.4&nbsp;add&nbsp;correct&nbsp;fasta&nbsp;files&nbsp;for&nbsp;fastp&nbsp;trimming<br>
1.7.3&nbsp;use&nbsp;bamtools&nbsp;more&nbsp;efficiently&nbsp;to&nbsp;filter&nbsp;mismatches&nbsp;(not&nbsp;adapt√≠ve&nbsp;to&nbsp;read&nbsp;length,&nbsp;but&nbsp;parameter&nbsp;enabled)<br>
1.7.2&nbsp;add&nbsp;ngmlr&nbsp;--min-residues&nbsp;cutoff,&nbsp;prefer&nbsp;to&nbsp;default&nbsp;0.25<br>
1.7.1&nbsp;add&nbsp;modified&nbsp;Nextera&nbsp;file&nbsp;and&nbsp;support&nbsp;for&nbsp;Trimmomatic&nbsp;trimming&nbsp;of&nbsp;Nextera&nbsp;adapters&nbsp;and&nbsp;transposase&nbsp;sequences<br>
1.7.0&nbsp;lint&nbsp;code&nbsp;with&nbsp;tool&nbsp;black<br>
1.6.9&nbsp;WIP&nbsp;-&nbsp;scale&nbsp;allowed&nbsp;number&nbsp;of&nbsp;allowed&nbsp;mismatches&nbsp;to&nbsp;read&nbsp;length,&nbsp;1&nbsp;every&nbsp;30bp&nbsp;?&nbsp;<br>
1.6.8&nbsp;remove&nbsp;--share&nbsp;from&nbsp;SLURM&nbsp;instructions&nbsp;(--share&nbsp;removed&nbsp;in&nbsp;modern&nbsp;2019&nbsp;SLURM)<br>
1.6.7&nbsp;add&nbsp;new&nbsp;viral&nbsp;ref&nbsp;EZV0_1_database2_cln.fasta<br>
1.6.6&nbsp;add&nbsp;new&nbsp;ref&nbsp;2020_03&nbsp;-&nbsp;same&nbsp;as&nbsp;2019_10,&nbsp;but&nbsp;removed&nbsp;Synthetic&nbsp;E.&nbsp;coli&nbsp;which&nbsp;collided&nbsp;with&nbsp;real&nbsp;E.&nbsp;coli&nbsp;when&nbsp;using&nbsp;mq30&nbsp;mode.<br>
1.6.5&nbsp;add&nbsp;new&nbsp;ref&nbsp;2019_10_meta_human_univec,&nbsp;improve&nbsp;helper&nbsp;scripts<br>
1.6.4&nbsp;solve&nbsp;bam.txt&nbsp;mq30&nbsp;problems<br>
1.6.3&nbsp;generalize&nbsp;conda&nbsp;to&nbsp;avoid&nbsp;specific&nbsp;filesystem<br>
1.6.2&nbsp;make&nbsp;more&nbsp;general&nbsp;for&nbsp;new&nbsp;users,&nbsp;improve&nbsp;initial&nbsp;error&nbsp;messages<br>
1.6.1&nbsp;solve&nbsp;ngmlr&nbsp;bugs,&nbsp;solve&nbsp;minimap2&nbsp;@SQ&nbsp;problem&nbsp;with&nbsp;--split-prefix&nbsp;temp_name<br>
1.6&nbsp;&nbsp;&nbsp;add&nbsp;ngmlr&nbsp;aligner,&nbsp;--longreads&nbsp;now&nbsp;omits&nbsp;Picard&nbsp;remove_dups&nbsp;by&nbsp;default&nbsp;(fails)<br>
1.5.1&nbsp;improve&nbsp;SOLiD&nbsp;adapter&nbsp;removal&nbsp;with&nbsp;fastp&nbsp;-&nbsp;configure&nbsp;var&nbsp;adapter_fastp<br>
1.5&nbsp;restructure&nbsp;wochenende_reporting,&nbsp;requires&nbsp;Python3.6+<br>
1.4&nbsp;add&nbsp;wochenende_plot.py&nbsp;file&nbsp;plotting<br>
1.3&nbsp;add&nbsp;samtools&nbsp;flagstat&nbsp;to&nbsp;get&nbsp;per&nbsp;cent&nbsp;reads&nbsp;aligned<br>
1.3&nbsp;add&nbsp;--testWochenende&nbsp;tests&nbsp;to&nbsp;test&nbsp;pipeline&nbsp;functionality&nbsp;and&nbsp;report&nbsp;success&nbsp;on&nbsp;a&nbsp;small&nbsp;reference<br>
1.2&nbsp;add&nbsp;--no-prinseq&nbsp;option&nbsp;to&nbsp;cancel&nbsp;prinseq&nbsp;read&nbsp;exclusion<br>
1.2&nbsp;add&nbsp;generation&nbsp;of&nbsp;unaligned&nbsp;reads&nbsp;(function&nbsp;runGetUnmappedReads)&nbsp;in&nbsp;FASTQ&nbsp;format<br>
1.0&nbsp;add&nbsp;reporting<br>
0.6&nbsp;add&nbsp;Minimap2<br>
0.5&nbsp;improve&nbsp;argument&nbsp;parsing&nbsp;and&nbsp;checks<br>
0.4&nbsp;use&nbsp;on&nbsp;genomics&nbsp;and&nbsp;metagenomics&nbsp;with&nbsp;awk&nbsp;filter&nbsp;script<br>
0.3&nbsp;add&nbsp;bwa-mem<br>
0.21&nbsp;add&nbsp;tool&nbsp;functions<br>
0.2&nbsp;add&nbsp;references<br>
0.1&nbsp;initial&nbsp;commits</tt></p>
<p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#aa55cc">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Modules</strong></big></font></td></tr>
    
<tr><td bgcolor="#aa55cc"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><table width="100%" summary="list"><tr><td width="25%" valign=top><a href="argparse.html">argparse</a><br>
<a href="os.html">os</a><br>
</td><td width="25%" valign=top><a href="shutil.html">shutil</a><br>
<a href="subprocess.html">subprocess</a><br>
</td><td width="25%" valign=top><a href="sys.html">sys</a><br>
<a href="time.html">time</a><br>
</td><td width="25%" valign=top></td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#eeaa77">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Functions</strong></big></font></td></tr>
    
<tr><td bgcolor="#eeaa77"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><dl><dt><a name="-abra"><strong>abra</strong></a>(stage_infile, fasta, threads)</dt></dl>
 <dl><dt><a name="-addToProgress"><strong>addToProgress</strong></a>(func_name, c_file)</dt></dl>
 <dl><dt><a name="-calmd"><strong>calmd</strong></a>(stage_infile, fasta)</dt></dl>
 <dl><dt><a name="-check_arguments"><strong>check_arguments</strong></a>(args)</dt></dl>
 <dl><dt><a name="-createProgressFile"><strong>createProgressFile</strong></a>(args)</dt></dl>
 <dl><dt><a name="-createTmpDir"><strong>createTmpDir</strong></a>(path_tmpdir)</dt></dl>
 <dl><dt><a name="-deriveRead2Name"><strong>deriveRead2Name</strong></a>(seRead)</dt></dl>
 <dl><dt><a name="-getAverageReadLengthBAM"><strong>getAverageReadLengthBAM</strong></a>(bamfile)</dt></dl>
 <dl><dt><a name="-main"><strong>main</strong></a>(args, sys_argv)</dt></dl>
 <dl><dt><a name="-markDups"><strong>markDups</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-markDupsSamtools"><strong>markDupsSamtools</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-rejigFiles"><strong>rejigFiles</strong></a>(stage, stage_infile, stage_outfile)</dt></dl>
 <dl><dt><a name="-runAfterQC"><strong>runAfterQC</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runAligner"><strong>runAligner</strong></a>(stage_infile, aligner, index, noThreads, readType)</dt></dl>
 <dl><dt><a name="-runBAMindex"><strong>runBAMindex</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runBAMsort"><strong>runBAMsort</strong></a>(stage_infile, readType)</dt></dl>
 <dl><dt><a name="-runBAMsortByName"><strong>runBAMsortByName</strong></a>(stage_infile, readType)</dt></dl>
 <dl><dt><a name="-runBamtools"><strong>runBamtools</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runBamtoolsAdaptive"><strong>runBamtoolsAdaptive</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runBamtoolsFixed"><strong>runBamtoolsFixed</strong></a>(stage_infile, numberMismatches)</dt></dl>
 <dl><dt><a name="-runEATrimming"><strong>runEATrimming</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runFastQC"><strong>runFastQC</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runFastUniq"><strong>runFastUniq</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runFastpPE"><strong>runFastpPE</strong></a>(stage_infile_1, stage_infile_2, noThreads, adapter_path)</dt></dl>
 <dl><dt><a name="-runFastpSE"><strong>runFastpSE</strong></a>(stage_infile, noThreads, adapter_path)</dt></dl>
 <dl><dt><a name="-runFixmate"><strong>runFixmate</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runFunc"><strong>runFunc</strong></a>(func_name, func, cF, newCurrentFile, *extraArgs)</dt><dd><tt>Used&nbsp;in&nbsp;the&nbsp;main&nbsp;function&nbsp;to&nbsp;compose&nbsp;the&nbsp;pipeline.&nbsp;Runs&nbsp;a&nbsp;function&nbsp;and&nbsp;adds&nbsp;<br>
it&nbsp;to&nbsp;the&nbsp;progress&nbsp;file.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;func_name&nbsp;(str):&nbsp;The&nbsp;function's&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;func&nbsp;(fun):&nbsp;The&nbsp;function&nbsp;to&nbsp;run<br>
&nbsp;&nbsp;&nbsp;&nbsp;cF&nbsp;(str):&nbsp;the&nbsp;current&nbsp;file&nbsp;to&nbsp;operate&nbsp;on<br>
&nbsp;&nbsp;&nbsp;&nbsp;*extraArgs:&nbsp;any&nbsp;additional&nbsp;arguments&nbsp;for&nbsp;the&nbsp;function<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;str:&nbsp;The&nbsp;new&nbsp;current&nbsp;file&nbsp;which&nbsp;is&nbsp;used&nbsp;for&nbsp;the&nbsp;next&nbsp;step.&nbsp;It&nbsp;is&nbsp;defined<br>
&nbsp;&nbsp;&nbsp;&nbsp;by&nbsp;the&nbsp;input&nbsp;function.</tt></dd></dl>
 <dl><dt><a name="-runGetUnmappedReads"><strong>runGetUnmappedReads</strong></a>(stage_infile, readType)</dt></dl>
 <dl><dt><a name="-runIDXstats"><strong>runIDXstats</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runMQ30"><strong>runMQ30</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runPerlDup"><strong>runPerlDup</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runPrinseq"><strong>runPrinseq</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runPrinseqPE"><strong>runPrinseqPE</strong></a>(stage_infile_1, stage_infile_2)</dt></dl>
 <dl><dt><a name="-runSamtoolsFlagstat"><strong>runSamtoolsFlagstat</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runStage"><strong>runStage</strong></a>(stage, programCommand)</dt><dd><tt>Run&nbsp;a&nbsp;stage&nbsp;of&nbsp;this&nbsp;Pipeline<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;stage&nbsp;(str):&nbsp;the&nbsp;stage's&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;programCommand&nbsp;(str):&nbsp;the&nbsp;command&nbsp;to&nbsp;execute&nbsp;as&nbsp;new&nbsp;subprocess</tt></dd></dl>
 <dl><dt><a name="-runTMTrimming"><strong>runTMTrimming</strong></a>(stage_infile, adapter_file)</dt></dl>
 <dl><dt><a name="-runTMTrimmingPE"><strong>runTMTrimmingPE</strong></a>(stage_infile, adapter_file)</dt></dl>
 <dl><dt><a name="-runTests"><strong>runTests</strong></a>(stage_infile)</dt></dl>
 <dl><dt><a name="-runTrimGalorePE"><strong>runTrimGalorePE</strong></a>(stage_infile, noThreads, adapter_path)</dt><dd><tt>####################&nbsp;TODO&nbsp;!!!!!!!!!!!!!!!!&nbsp;Have&nbsp;never&nbsp;done&nbsp;this&nbsp;for&nbsp;TrimGalore&nbsp;AND&nbsp;how&nbsp;does&nbsp;it&nbsp;do&nbsp;PE&nbsp;output?</tt></dd></dl>
 <dl><dt><a name="-runTrimGaloreSE"><strong>runTrimGaloreSE</strong></a>(stage_infile, noThreads, nextera)</dt></dl>
</td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#55aa55">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Data</strong></big></font></td></tr>
    
<tr><td bgcolor="#55aa55"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><strong>IOthreadsConstant</strong> = '8'<br>
<strong>adapter_fastp_general</strong> = '/lager2/rcug/seqres/contaminants/2020_02/adapters/adapters.fa'<br>
<strong>adapter_fastp_nextera</strong> = '/lager2/rcug/seqres/contaminants/2020_02/adapters/NexteraPE-PE.fa'<br>
<strong>adapter_fastp_solid</strong> = '/lager2/rcug/seqres/contaminants/2020_02/adapters/adapters_solid.fa'<br>
<strong>adapter_nextera</strong> = '/lager2/rcug/seqres/contaminants/2020_02/adapters/NexteraPE-PE.fa'<br>
<strong>adapter_truseq</strong> = '/mnt/ngsnfs/tools/miniconda3/envs/wochenende/share/trimmomatic-0.38-0/adapters/TruSeq3-PE.fa'<br>
<strong>ea_adapter_fasta</strong> = '/lager2/rcug/seqres/contaminants/2020_02/adapters/adapters.fa'<br>
<strong>fileList</strong> = []<br>
<strong>path_abra_jar</strong> = '/mnt/ngsnfs/tools/abra2/abra2_latest.jar'<br>
<strong>path_afterqc</strong> = '/mnt/ngsnfs/tools/afterQC/AfterQC-0.9.6/after.py'<br>
<strong>path_bamtools</strong> = 'bamtools'<br>
<strong>path_bwa</strong> = 'bwa'<br>
<strong>path_fastp</strong> = 'fastp'<br>
<strong>path_fastq_mcf</strong> = 'fastq_mcf'<br>
<strong>path_fastqc</strong> = 'fastqc'<br>
<strong>path_fastuniq</strong> = 'fastuniq'<br>
<strong>path_java</strong> = 'java'<br>
<strong>path_minimap2</strong> = 'minimap2'<br>
<strong>path_ngmlr</strong> = 'ngmlr'<br>
<strong>path_perl</strong> = 'perl'<br>
<strong>path_perldup</strong> = '/mnt/ngsnfs/tools/Wochenende/dependencies/remove_pcr_duplicates.pl'<br>
<strong>path_prinseq</strong> = 'prinseq-lite.pl'<br>
<strong>path_refseq_dict</strong> = {'2016_06_1p_genus': '/working2/tuem/metagen/refs/2016/bwa/2016_06_PPKC_metagenome_test_1p_genus.fa', '2016_06_1p_spec': '/working2/tuem/metagen/refs/2016/bwa/2016_06_PPKC_metagenome_test_1p_spec_change.fa', '2016_06_1p_spec_corrected': '/lager2/rcug/seqres/metagenref/bwa/2016_06_PPKC_metagenome_test_1p_spec_change_cln.fa', '2019_01_meta': '/lager2/rcug/seqres/metagenref/bwa/all_kingdoms_refseq_2019_Jan_final.fasta', '2019_01_meta_mouse': '/lager2/rcug/seqres/metagenref/bwa/all_kingdoms_refseq_2019_Jan_final_mm10_no_human.fasta', '2019_01_meta_mouse_ASF': '/lager2/rcug/seqres/metagenref/bwa/mm10_plus_ASF.fasta', '2019_01_meta_mouse_ASF_OMM': '/lager2/rcug/seqres/metagenref/bwa/mm10_plus_ASF_OMM.fasta', '2019_01_meta_mouse_OMM': '/lager2/rcug/seqres/metagenref/bwa/mm10_plus_OMM.fasta', '2019_10_meta_human': '/lager2/rcug/seqres/metagenref/bwa/refSeqs_allKingdoms_201910_3.fasta', '2019_10_meta_human_univec': '/lager2/rcug/seqres/metagenref/bwa/refSeqs_allKingdoms_201910_3_with_UniVec.fasta', ...}<br>
<strong>path_sambamba</strong> = 'sambamba'<br>
<strong>path_samtools</strong> = 'samtools'<br>
<strong>path_tmpdir</strong> = '/ngsssd1/rcug/tmp/'<br>
<strong>path_trim_galore</strong> = 'trim_galore'<br>
<strong>path_trimmomatic</strong> = 'trimmomatic'<br>
<strong>stage_infile</strong> = ''<br>
<strong>stage_outfile</strong> = ''<br>
<strong>trim_galore_min_quality</strong> = '20'<br>
<strong>version</strong> = '1.8.3 - January 2021'</td></tr></table>
</body></html>